### ISSUE TITLE
Differential paths coverage between fuzzers

### ISSUE THREAD

**Issue by inferno-chromium** [google/fuzzbench#686]
@eqv's idea
Either
1: Capture all inputs generated during fuzzing and then store path/state transitions that happen.
2. Capture path/state transitions in real-time (without overslowing things).

**Comment by eqv** [comment]
Just to avoid flaming: The idea is to sample a very small percentage of the inputs during fuzzing.

**Comment by inferno-chromium** [comment]
> Just to avoid flaming: The idea is to sample a very small percentage of the inputs during fuzzing.

Corrected in description.

**Comment by mboehme** [comment]
What are path or state transitions?

**Comment by lszekeres** [comment]
> What are path or state transitions?

I think that refers to a "coverage point" (eg. edge / block / [region](https://clang.llvm.org/docs/SourceBasedCodeCoverage.html) / or what have you).

**Comment by mboehme** [comment]
Great idea! I suppose this would show the coverage overlap (difference) between fuzzers? Might help with debugging a fuzzer.

**Comment by eqv** [comment]
exactly & relative frequency of the areas explored.

**Comment by lszekeres** [comment]
> Great idea! I suppose this would show the coverage overlap (difference) between fuzzers? Might help with debugging a fuzzer.

Note that the coverage overlap/difference between fuzzers is already shown by the [unique coverage plots](https://www.fuzzbench.com/reports/2020-09-07/index.html).

**Comment by inferno-chromium** [comment]
I think this one is about edge/region transitions and not about edges/regions covered.

**Comment by aflgo** [comment]
Maybe @eqv can clarify, but I understood it more like what he mentioned in his blog post.

![](https://hexgolems.com/imgs/fuzzeval2.png)

https://hexgolems.com/2020/08/on-measuring-and-visualizing-fuzzer-performance/

(I would also be intetested in the data).

**Comment by eqv** [comment]
Like @aflgo said, initially it was meant to be about the difference in frequency. Which is stronger than difference in coverage. The downside is that it requires additional changes to be measured. Note that we cannot simply compute this based on the queue. Instead we would have to live sample from the fuzzing process. Like the visualization done by Brandon, but as a proper measurement: https://twitter.com/gamozolabs/status/1288440436680306688

**Comment by eqv** [comment]
For the record: While super cool to watch, I don't think that kind of visualization (cookie dough) will help to understand the performance of fuzzers on real programs - We would need better ways to aggregate and plot the data.

**Comment by mboehme** [comment]
Fuzzbench distinguishes runner and measurer and only the seed corpus ever reaches the measurer. So, getting coverage statistics right from the fuzzing process might be a bit more difficult.

If covered_regions.json contains information on the number of seeds that exercise a certain region, you could generate at least an approximation. Does that work?

**Comment by eqv** [comment]
It's certainly nontrivial to predict this from the queue... you will at least also need how often the seed where mutated, and even then you will probably only get a very poor approximation (because most mutations will go to an early error path that is hence triggered MUCH more often that the part that was actually the reason why the seed was stored). Would be interesting to compare though.

**Comment by mboehme** [comment]
The following is based on data from [Stephan Lipp](https://www.in.tum.de/i04/lipp/) (@sphl) of TU Munich. I used that data to play around with ways to visualize coverage progress. The graphs can be auto-generated from a CSV file containing function-level coverage information over time (amongst others) and the CG dot file produced by LLVM, plus a bit of R and GraphViz.

Fun little coverage visualizations inspired by @Gamozolabs [cookie-dough](https://github.com/gamozolabs/cookie_dough). These are call graphs of freetype2. Red is uncovered in all trials, yellow is covered by at least one trial and less than 20 trials, Green is covered in all 20 trials. 

Still playing around with better ways to visualize. Was hoping to see some kind of coverage-frontier advancing forward from the root node. Instead it's looks like increasingly more parts of a Christmas tree coming on.

![freetype2_naeto_small](https://user-images.githubusercontent.com/5443180/94344979-4e953e80-0066-11eb-8d60-780826be6a7f.gif)


|AFL (23 hours in 32 frames)|Honggfuzz (23 hours in 32 frames)|
|--|--|
|![Honggfuzz](https://user-images.githubusercontent.com/5443180/94215941-6c0cc000-ff21-11ea-89ad-0f396f8fc298.gif)|![AFL](https://user-images.githubusercontent.com/5443180/94215859-2bad4200-ff21-11ea-9070-9e765ed04b7e.gif)|

## Both (CG as a Tree)
*Maybe this should differential coverage instead?*

![freetype2](https://user-images.githubusercontent.com/5443180/94343657-f0b02900-005c-11eb-96b5-75accabebdb3.gif)
Graphviz doesn't let me break the ranks to produce a format with less blank space.

## Graphs
At the top, the number of BBs over the number of trials covering the BBs at the end of the 24hr campaign. For instance, for AFL a bit over 10k BBs are covered in none of the trials while a bit over 3k BBs are covered in all trials. There are only 2 BBs that are covered by 9 and 19 trails, respectively.

At the bottom, we see coverage over time for a single run. Obviously, Honggfuzz continuous to make progress while the campaign for AFL is stalled.

*Maybe this one should also be plotted over time?*
![image](https://user-images.githubusercontent.com/5443180/94344669-30c6da00-0064-11eb-9d60-f6e4fe418b94.png)

**Comment by Danushka-liyanage** [comment]
> Maybe @eqv can clarify, but I understood it more like what he mentioned in his blog post.
> 
> ![](https://camo.githubusercontent.com/2b83fe8228a5f08bc48873fe21eb1a3a1a4b9180/68747470733a2f2f686578676f6c656d732e636f6d2f696d67732f66757a7a6576616c322e706e67)
> 
> https://hexgolems.com/2020/08/on-measuring-and-visualizing-fuzzer-performance/
> 
> (I would also be intetested in the data).

We considered two fuzzing executions, AFL and Honggfuzz on freetype2. Similar to this graph, the differences between hit counts of AFL and Hinggfuzz for each basic block are plotted. Also, the hit counts are represented on the same graph. The blue and green color dots represent hit_counts(AFL) and -hit_counts(Honggfuzz) for each basic block. Make sure that we have excluded all the basic blocks which had 0 hits counts from both the fuzzers.
![image](https://user-images.githubusercontent.com/57780422/94516626-8cae8000-0269-11eb-91fb-0baa764b1dc6.png)

**Comment by mboehme** [comment]
Thanks Danushka! Maybe I can add an interpretation. 

We use the log10(hits) just because the scale is substantial. The X axis shows the basic block IDs ordered by log10(AFL_hits) - log10(Hongg_hits). The blue dots are log10(AFL_hits) and the green dots are -log10(Hongg_hits).

On the left, we can see BBs that AFL excutes about 1 order of magnitude more often. On the right, we can see BBs that Hongg executes about 10 order of magnitudes more often. In this example, Honggfuzz clearly executes more BBs and the majority of BBs several orders of magnitude more often than AFL.

**Comment by eqv** [comment]
Sweet! This is exactly the kind of plot I was thinking of! The differences between hongg fuzz and AFL are pretty big (in terms of coverage reached overall) but I think this neatly shows how almost all of the basic blocks where AFL is doing better than honggfuzz are also basic blocks that honggfuzz explores very thoroughly already. Are the datapoints sampled over the whole campaign or are they calculated from the queue?

**Comment by mboehme** [comment]
That's from the coverage profile generated by executing the entire queue at the end of the campaign.

**Comment by eqv** [comment]
hmm that means that early basic blocks will be under represented, as they are found in a few inputs, but where ran a lot of times. And basic blocks that were found later are over represented, as they were executed far less than the early ones, even though they both appear the same # times in the queue. Right?

**Comment by eqv** [comment]
Is there a way to access this data? I wonder if it would be helpful to display how often the losing fuzzer covered the same basic block. So you can easily tell the difference between 30_000 vs 40_0000 and 0 vs 10_000.

**Comment by mboehme** [comment]
> Is there a way to access this data?

We got the data from @sphl. Not sure, we can share. PR #778 is to generate that data in Fuzzbench.

> I wonder if it would be helpful to display how often the losing fuzzer covered the same basic block. So you can easily tell the difference between 30_000 vs 40_0000 and 0 vs 10_000.

This should already be visible. For any BB where the line is below 0 (i.e., AFL is the losing fuzzer), blue indicates log10(hits + 1) for AFL.

**Comment by Danushka-liyanage** [comment]
> Is there a way to access this data? I wonder if it would be helpful to display how often the losing fuzzer covered the same basic block. So you can easily tell the difference between 30_000 vs 40_0000 and 0 vs 10_000.

I think this would help. Say ntrials is the number of times a given BB hit at least once out of k runs. We plotted the ntrials(AFL)-ntrials(Hongg) along with ntrials(AFL) and -ntrails(Hongg) in the same graph. In data, there were k=20 runs for each fuzzer-subject combination. Same as above, blue and green dots represent AFL and Honggfuzz trials respectively. 
[Trial_counts.pdf](https://github.com/google/fuzzbench/files/5302593/Trial_counts.pdf)

**Comment by eqv** [comment]
> This should already be visible. For any BB where the line is below 0 (i.e., AFL is the losing fuzzer), blue indicates log10(hits + 1) for AFL.

Yes, in theory that's true, but it's really hard to see - in general it seems like most basic blocks where AFL is better are also found by Honggfuzz many times, but imho we can't easily see if there is some basic blocks that are found significantly fewer times.

**Comment by mboehme** [comment]
That's log10(hits) sorted by Honggfuzz hits.
![image](https://user-images.githubusercontent.com/5443180/94750904-884da880-03ca-11eb-91e0-f75c72b32024.png)
